### PR TITLE
feat: add API Center Plugin for GitHub Copilot for Azure

### PR DESCRIPTION
**Pull request by formulahendry**
demo video: https://microsoftapc-my.sharepoint.com/:v:/g/personal/junhan_microsoft_com1/EQk-cMuWJlFJn-qKcb_ljXEBl3Npd-K7SCjvOFA4Oli8VQ?e=oaYbmj&nav=eyJyZWZlcnJhbEluZm8iOnsicmVmZXJyYWxBcHAiOiJTdHJlYW1XZWJBcHAiLCJyZWZlcnJhbFZpZXciOiJTaGFyZURpYWxvZy1MaW5rIiwicmVmZXJyYWxBcHBQbGF0Zm9ybSI6IldlYiIsInJlZmVycmFsTW9kZSI6InZpZXcifX0%3D

### CODE REVIEW

File: src/copilot/generateOpenApi.ts [hunk 1]
```diff
@@ -0,0 +1,191 @@
...
+        } else {
+            return spectralDefaultRuleDescriptions;
+        }
+    }
+
+    async function getRulesetFileContent(rulesetFile: string): Promise<string | undefined> {
+        try {
+            let rulesetFileContent = "";
+            if (rulesetFile.startsWith('http://') || rulesetFile.startsWith('https://')) {
+                rulesetFileContent = await axios.get(rulesetFile).then(response => response.data);
```
Comment (wenytang-ms) [inline src/copilot/generateOpenApi.ts:137]: It's mixing await/async and promise resolution with .then()/.catch(), so just pick one.

File: package.json [hunk 2]
```diff
@@ -561,5 +561,9 @@
     "*.{js,jsx,css,ts,tsx}": [
       "npx eslint --cache --fix --quiet"
     ]
+  },
+  "agentMetadata": {
+    "version": "2.0",
```
Comment (adashen) [inline package.json:566]: fix version or match whole minor?

File: src/uiStrings.ts [hunk 3]
```diff
@@ -132,5 +132,14 @@ export class UiStrings {
     static readonly GetTreeView = vscode.l10n.t("Please connect to Azure API Center Service first.");
     static readonly APIControlPlaneView = vscode.l10n.t("API Center Management Plane");
     static readonly APIDataPlaneView = vscode.l10n.t("API Center Data Plane");
-    static readonly APIDataPlaneWiki = vscode.l10n.t("API Catalog Guide")
+    static readonly APIDataPlaneWiki = vscode.l10n.t("API Catalog Guide");
+    static readonly GenerateOpenApiProgress = vscode.l10n.t("Invoking Azure API Center to generate OpenAPI specification documentation...");
```
Comment (adashen) [inline src/uiStrings.ts:136]: Is it Azure API Center or VSCode Azure API Center Extension?

File: src/copilot/generateOpenApi.ts [hunk 4]
```diff
@@ -0,0 +1,191 @@
...
+        const chatResponseParts: vscode.ChatResponsePart[] = [];
+
+        chatResponseParts.push(new vscode.ChatResponseMarkdownPart(UiStrings.GenerateOpenApiRegisterApiDesc));
+        chatResponseParts.push(new vscode.ChatResponseCommandButtonPart({
+            title: UiStrings.GenerateOpenApiRegisterApiButton,
+            command: "azure-api-center.registerApi",
+            arguments: []
+        }));
+
+        if (rulesetFile) {
```
Comment (adashen) [inline src/copilot/generateOpenApi.ts:82]: if the fetching content from the rulesetFile failed, we still use this block of logic?

File: package.json (hunk 2 above)
Comment (formulahendry) [reply package.json:566]: This version string is defined by Azure Agent. We just follow it.

File: src/uiStrings.ts (hunk 3 above)
Comment (formulahendry) [reply src/uiStrings.ts]: Good point. I changed to `Azure API Center plugin`

File: src/copilot/generateOpenApi.ts (hunk 4 above)
Comment (formulahendry) [reply src/copilot/generateOpenApi.ts:82]: In `else` logic, it is mainly for users who have never set the API style guide. So even when reading rulesetFile failed, it is OK to go to this code block.

File: src/uiStrings.ts [hunk 5]
```diff
@@ -132,5 +132,14 @@ export class UiStrings {
...
     static readonly APIDataPlaneView = vscode.l10n.t("API Center Data Plane");
-    static readonly APIDataPlaneWiki = vscode.l10n.t("API Catalog Guide")
+    static readonly APIDataPlaneWiki = vscode.l10n.t("API Catalog Guide");
+    static readonly GenerateOpenApiProgress = vscode.l10n.t("Invoking Azure API Center plugin to generate OpenAPI specification documentation...");
+    static readonly GenerateOpenApiRegisterApiDesc = vscode.l10n.t("\n\nAzure API Center is a centralized hub for discovering, reusing, and governing all of your organization's APIs. Would you like to register your API in Azure API Center?");
+    static readonly GenerateOpenApiRegisterApiButton = vscode.l10n.t("Register your API in API Center");
+    static readonly GenerateOpenApiViewApiDesc = vscode.l10n.t("\n\nWould you like to see a list of all the APIs available in the Azure API Center?");
+    static readonly GenerateOpenApiViewApiButton = vscode.l10n.t("Show Azure API Center");
+    static readonly GenerateOpenApiSetRuleDesc = vscode.l10n.t("\n\nWould you like to ensure your API specification meets your organizational compliance guidelines?");
```
Comment (juliajuju93) [inline src/uiStrings.ts:141]: Can we change that string to:

"\n\nWould you like to analyze your API definitions for compliance with an API style guide?")

File: src/uiStrings.ts [hunk 6]
```diff
@@ -132,5 +132,14 @@ export class UiStrings {
     static readonly GetTreeView = vscode.l10n.t("Please connect to Azure API Center Service first.");
     static readonly APIControlPlaneView = vscode.l10n.t("API Center Management Plane");
     static readonly APIDataPlaneView = vscode.l10n.t("API Center Data Plane");
-    static readonly APIDataPlaneWiki = vscode.l10n.t("API Catalog Guide")
+    static readonly APIDataPlaneWiki = vscode.l10n.t("API Catalog Guide");
+    static readonly GenerateOpenApiProgress = vscode.l10n.t("Invoking Azure API Center plugin to generate OpenAPI specification documentation...");
+    static readonly GenerateOpenApiRegisterApiDesc = vscode.l10n.t("\n\nAzure API Center is a centralized hub for discovering, reusing, and governing all of your organization's APIs. Would you like to register your API in Azure API Center?");
+    static readonly GenerateOpenApiRegisterApiButton = vscode.l10n.t("Register your API in API Center");
+    static readonly GenerateOpenApiViewApiDesc = vscode.l10n.t("\n\nWould you like to see a list of all the APIs available in the Azure API Center?");
```
Comment (juliajuju93) [inline src/uiStrings.ts:139]: Please change to:
("\n\nWould you like to see a list of all the APIs available in your API Center?");

File: src/uiStrings.ts [hunk 7]
```diff
@@ -132,5 +132,14 @@ export class UiStrings {
...
+    static readonly GenerateOpenApiProgress = vscode.l10n.t("Invoking Azure API Center plugin to generate OpenAPI specification documentation...");
+    static readonly GenerateOpenApiRegisterApiDesc = vscode.l10n.t("\n\nAzure API Center is a centralized hub for discovering, reusing, and governing all of your organization's APIs. Would you like to register your API in Azure API Center?");
+    static readonly GenerateOpenApiRegisterApiButton = vscode.l10n.t("Register your API in API Center");
+    static readonly GenerateOpenApiViewApiDesc = vscode.l10n.t("\n\nWould you like to see a list of all the APIs available in the Azure API Center?");
+    static readonly GenerateOpenApiViewApiButton = vscode.l10n.t("Show Azure API Center");
+    static readonly GenerateOpenApiSetRuleDesc = vscode.l10n.t("\n\nWould you like to ensure your API specification meets your organizational compliance guidelines?");
+    static readonly GenerateOpenApiSetRuleButton = vscode.l10n.t("Set active API style guide");
+    static readonly GenerateOpenApiRegenerateDesc = vscode.l10n.t("\n\nAfter API style guide is set, you could regenerate OpenAPI spec again.");
```
Comment (juliajuju93) [inline src/uiStrings.ts:143]: ("\n\nAfter your API style guide is set; you could regenerate the API specification again.")

File: src/uiStrings.ts [hunk 8]
```diff
@@ -132,5 +132,14 @@ export class UiStrings {
...
+    static readonly GenerateOpenApiRegenerateButton = vscode.l10n.t("Regenerate OpenAPI spec");
```
Comment (juliajuju93) [inline src/uiStrings.ts:144]: Regenerate API specification

File: src/copilot/getPlugins.ts [hunk 9]
```diff
@@ -0,0 +1,69 @@
...
+        const pluginRequest = args.localPluginRequest;
+        const eventName = `azure-api-center.agent.${pluginRequest.functionName}`;
+        let parsedError: IParsedError | undefined;
+
+        try {
+            TelemetryClient.sendEvent(`${eventName}.start`);
+            if (pluginRequest.functionName === genOpenApiFunctionName) {
+                return await GenerateOpenApi.handleGenerateOpenApi(args.agentRequest);
+            } else {
+                return { result: "Error: Function not found." };
```
Comment (wenytang-ms) [inline src/copilot/getPlugins.ts:41]: do we need to handle this error in telemetry?

Comment (formulahendry) [reply src/copilot/getPlugins.ts:41]: Function name will be included in event name - `azure-api-center.agent.${pluginRequest.functionName}`, so no need to add extra telemetry for this.

**Review by wenytang-ms — APPROVED** [review]

**Pull request merged**